Fix #10664 by fixing `Ctype.copy` in  `erase_either` mode

garrigue opened the pull request:
I proposed a first fix of #10664 by restricting `generalize_structure` in #11579.

However, a more careful analysis shows that the original source of the problem is not in `generalize_structure`, but in that `Ctype.copy` was keeping a `Tnil` row after erasing some `Reither` fields.
After the fix, we make sure that we always use a fresh variable in that case, since we are creating a partial copy, which will be unified with the real type later.

Compared to #11579, this just fixes the bug without introducing incompatibilities.
The question of the correctness of `generalize_structure` is still open (it seems to rely somehow on how we are using it).

garrigue committed:
Fix #10664

garrigue commented:
Sorry, this does not work. Please wait a bit more.

garrigue committed:
Fix real cause of #10664 in copy_sep

garrigue commented:
OK, I found the real cause.
Actually, I was scratching my head, seeing types such as
```
         exp_type =
          {id=1606;level=2;scope=0;desc=
           {row_fields=
            [B, RFeither(true,[],false,ref(RFpresent None));
             A, RFeither(true,[],false,ref(RFpresent None))];
            row_more={id=1611;level=2;scope=0;desc=Tnil}; row_closed=true;
            row_fixed=None; row_name=None}};
```
This is not supposed to happen: `Tnil` is only used for static rows, not for those with remaining structural polymorphism.

It turns out that the culprit was `copy_sep`.
When we copy a polymorphic variant that contains universal type variables, either as row variable or in a field, we work in unsharing mode, i.e. duplicating everything until we reach either a recursion, or a node with no universal variables underneath, and switch back to sharing mode.
The trouble is that we could end up unsharing non-static polymorphic variant type, i.e. turning
```
[< `A | `B] as 'a -> 'a
```
into
```
[< `A | `B] -> [< `A | `B]
```
More precisely, a shared row variable was kept, but the presence information of the fields was unshared.
Applying such a function to
```
[`A | `B]
```
resulted in the above strange internal type.

This is fixed by keeping a map of copied non-static polymorphic variants inside `copy_sep`. On the first occurence we still copy in unsharing mode, but when we meet it again somewhere else we keep the sharing.
I chose to keep the fix to `copy` too, as it makes the code more readable anyway.

BTW, `generalize_structure` appears to be OK. When we copy a non-static polymorphic variant whose row-variable is not generalized, we use the `keep` mode of `copy_row`, which shares the extension variables in each field.  So at least it should be sound. There may still be problems with levels: in theory we're supposed to ensure that anything that goes into those extension variables has a level lower or equal to the row variable, but this is easy to miss. Next step of the abstraction work?

garrigue commented:
Reminder: this is a soundness bug.

Also, sorry for the confusion if there was some: the fix is complete, it just changed in the middle of the branch, after a better analysis.

gasche commented:
Naive question: what does "without sharing" mean and why do we want that?

My vague understanding is that this function is in charge of substituting universally-bound variables (univars) away, at polymorphic-instantiation time, to replace them by normal variables: `'a. 'a -> int * bool` becomes `'b -> int * bool` for a fresh flexible variable `´b`.

I think that when the function reaches subexpressions that are univar-closed, it is intended substitution becomes the identity: univar-closed subtypes are shared. This is what the comment says, but I don't understand the relevant code:

```ocaml
  if is_Tvar ty || may_share && TypeSet.is_empty univars then
    if get_level ty <> generic_level then ty else
    let t = newstub ~scope:(get_scope ty) in
    delayed_copy :=
      lazy (Transient_expr.set_stub_desc t (Tlink (copy cleanup_scope ty)))
      :: !delayed_copy;
    t
```

(this branch is taken for flexible variables and for univar-closed subexpressions) `delayed_copy` looks like it will (later? why are we delaying) call `copy cleanup_scope ty`, so it looks like in fact each time we encounter this node `ty` in the type graph we will perform an independent copy, turning eg. `'b. (int * bool) as 'a -> 'b * 'a` into `(int * bool) -> 'c * (int * bool)`, with the `(int * bool)` nodes unshared. (I guess this is wrong because it contradicts the documentation, so I suppose there is in fact some sharing here, how?)

Then the case for `Tvariant` and `Tfield` both try to handle their row variables in a special way, but in a different special way. For `Tfield`, we call ourselves recursively on the row with `~may_share:false`, which disables the special sharing-of-univars-closed-subexpressions that I didn't understand above. We do the same in the `Tvariant` case, but the new PR in facts adds an extra level of caching on top that seems designed to still partly share those row variables.

garrigue commented:
"Without sharing" means that we copy as though the type were a (regular) tree rather than a graph, until we reach a node with no free universal variables. In that last case we copy it using the normal copying function, which copies as a graph. This copy is delayed, because we want to copy all the remaining parts simultaneously, to keep the sharing.
I don't have all the examples at hand, but IIRC the main reason we need to copy without sharing is that the same node may be accessed from different paths, and need to be copied separately because the free variables are different.

The typical example is as follows:
```ocaml
let o = object (self : 's) method m : 'a. 'a -> 's = fun x -> self end;;
val o : < m : 'a. 'a -> 'b > as 'b = <obj>
(o#m true, o#m 1);;
- : (< m : 'a. 'a -> 'b > as 'b) * (< m : 'a. 'a -> 'c > as 'c) = (<obj>, <obj>)
```
Here `o#m` returns a function from anything to a copy of the type of `o`, and you can see that we get two distinct copies here.

The `~may_share:false` is intended to prevent switching from unsharing mode to normal mode at an internal node (not visible in the printed type). The new behavior for polymorphic variants ensures that we share non-static polymorphic variant types between different branches, as their field kinds cannotbe shared individually. Note that kinds in object types work differently, and can be shared even without sharing the method type, so we can share them always (which makes sense as object types are not kind-polymorphic anyway).

gasche commented:
With the remark of Jacques I can answer my own question about the paragraph I quoted above.

```ocaml
  if is_Tvar ty || may_share && TypeSet.is_empty univars then
    if get_level ty <> generic_level then ty else
    let t = newstub ~scope:(get_scope ty) in
    delayed_copy :=
      lazy (Transient_expr.set_stub_desc t (Tlink (copy cleanup_scope ty)))
      :: !delayed_copy;
    t
```

It looks like this code may perform several independent copies of `t` if it is encountered several times during the graph traversal, but in fact:
1. `cleanup_scope`, which would be easier to read if named `copy_scope`, shares mutable state about the copying operation, so several calls to `copy` with the same `copy_scope` (initialized by `For_copy.with_scope`, in this case in `instance_poly`) can be thought of as a single copy on multiple subexpressions.
2. `copy` works by mutating its argument in-place with an undo log on the side (in `copy_scope`), so it would not be safe to call `copy copy_scope ty` directly -- it would interfere with the `copy_sep` logic by mutating its input under its feet. This is a reason to delay the `copy` call to the end of the `copy_sep` processing, using `lazy`.

@garrigue sorry for the naive questions again, but I'm not sure I follow your example. I can see that the self types `< ... > as 'b` and `< ... > as 'c` are not shared in the result, but there are two occurrences of `o#m` in the source program so (I think?) we are performing two different instantiations (in two different copy scopes). I would have expected an example where a single call to `copy_sep` unshares types (for example copying starting from `'a -> 'b * 'b` in `< m : 'a . 'a -> 'b * 'b > as 'b` and getting `'a -> < m : 'a . 'a -> 'b * 'b> * < m : 'a . 'a -> 'b * 'b >). But here you have an example of two distinct `copy_sep` calls returning non-shared results, right? So when you talk about sharing, you mean sharing type nodes between in the result of two different calls to the same function, instead of sharing type nodes that are copies of the same input node in the same call?

garrigue commented:
One can also write an example like that:
```ocaml
let f (x : < m : 'a. (< n : 'a; p : 'b > as 'b) * 'b >) = x#m;;
val f :
  < m : 'a. (< n : 'a; p : 'b > as 'b) * 'b > ->
  (< n : 'd; p : 'c > as 'c) * (< n : 'd; p : 'e > as 'e) = <fun>
```
which differs from
```ocaml
let g (x : (< n : 'a; p : 'b > as 'b) * 'b) = x;;
val g : (< n : 'a; p : 'b > as 'b) * 'b -> 'b * 'b = <fun>
```
This said, the last example in `-principal` mode already gives
```ocaml
val g :
  (< n : 'a; p : 'b > as 'b) * (< n : 'a; p : 'c > as 'c) ->
  (< n : 'a; p : 'd > as 'd) * (< n : 'a; p : 'e > as 'e) = <fun>
```
due to the different behavior of `as` in that case, so the width aspect of the unsharing is less important than its depth aspect, i.e. calling a polymorphic method whose type is recursive unrolls its type, duplicating until the start of the recursion is reached.
```ocaml
let h (x : < m : 'a. < n : 'a; p : 'b > > as 'b) = x#m;;
val h :
  (< m : 'a. < n : 'a; p : 'b > > as 'b) ->
  < n : 'c; p : < m : 'a. < n : 'a; p : 'd > > as 'd > = <fun>
```

gasche commented:
In your last example, I can only observe this de-duplication-in-depth behavior in `-principal` mode (on my 4.14 switch), not in non-principal mode. Is this really an example of `copy_sep` at work or something else?

```ocaml
# #principal false;;
# let h (x : < m : 'a. < n : 'a; p : 'b > > as 'b) = x#m;;
val h : (< m : 'a. < n : 'a; p : 'b > > as 'b) -> < n : 'c; p : 'b > = <fun>
# #principal true;;
# let h (x : < m : 'a. < n : 'a; p : 'b > > as 'b) = x#m;;
val h :
  (< m : 'a. < n : 'a; p : 'b > > as 'b) ->
  < n : 'c; p : < m : 'a. < n : 'a; p : 'd > > as 'd > = <fun>
```

Taking a step back: it still don't understand what is the motivation for `copy_sep` (besides substituting some universal variables away, which `copy` itself does not do), for its unsharing aspect. I wish the function came with a comment that explains why it is there by showing a simple example program that would be type-checked incorrectly without its unsharing logic, and explain why without-unsharing result would be wrong. I currently don't understand what @garrigue writes well enough to offer such an example.

garrigue commented:
@gasche My explanation may have been confusing. I was not talking of unsharing wrt the original type, but of unsharing inside the type. Here the point is that `< n : 'a; p : 'b >` is copied separately because `'a` is replaced by a normal type variable outside, but kept as a universal type variable inside. Physical sharing of the inside type, which is copied using `copy` rather than `copy_sep`, depends indeed on `-principal`.
It is clear that `copy_sep` is needed, i.e. some nodes need to be instantiated differently depending on the context, but how much unsharing is really needed, and whether the criterion in`copy_sep` is minimal, this is a different question. I wrote this code 20 years ago, and I'm afraid I didn't keep all the examples around. Many of them are in `typing-poly/poly.ml`, but I'm not sure it is really exhaustive.
I think that the original principle behind `copy_sep` was "unshare anything as long as there are free universal type variables inside". But of course this cannot be done since a recursive type may also contain free universal type variables, like in `f` above, so we also need to close this recursion, and at the right place, hence the somehow convoluted criterion, which may be overly cautious. There is no formal proof that this is correct, but this seems to make sense, except for the lack of sharing uncovered in #10664.

garrigue committed:
add more examples

garrigue commented:
Rebased.
To be more precise about the criterion, the code could be improved.
First, there is a check `is_Tunivar ty`, but `ty` is supposed to be in `visited`, and if you look below you see that only `Tarrow` ... `Tpackage` are supposed to be added to `visited`, so this check should always return false.
For the criterion itself, we check that none of the universal variables were captured since the last occurrence. This part seems minimal.

I have added extra examples to see how polymorphic variants and polymorphic methods interact. If the row variable is explicitly captured as a universal variable, the row is static, and correctly duplicated. If the row variable is not captured, then some universal variables may leak through, which seems unavoidable. In one case the leakage is not detected immediately. I will look at why this happens, but this is detected later on through unification.

garrigue committed:
hygiene

garrigue committed:
also fix Tvar case; no assumption on opening

garrigue committed:
oops

garrigue commented:
I have fix the leakage problem. This causes a few changes in typing-gadts and typing-poly, but this actually improves error message.
I have also added extra code to make the behavior more correct when the row-variable is `Tunivar`.
However, this does not seem necessary. I will make an extra pass on the code expliciting the invariant that the only free univars are the outermost ones.

gasche commented:
(I understood things better again, thanks for the extra explanations. Below is an attempt to reformulate, explain why "it is clear that `copy_sep` is needed".)

In your example `< m : 'a. < n : 'a; p : 'p > > as 'p`, we are instantiating  `x#m`, so the type that we are copying is the body of the Tpoly node `'a . ...`, which can be written `(< n : 'a ; p : < m : 'a . 'b > >) as 'b`, with a free universal variable `'a` that is bound above the recursive inner occurrence of `'b` (this type only makes sense with a nominal representation of universal variables, not if we view them as abstract binders). We want to instantiate the outer universal variable `'a`, so the result of copying the outer occurrence of `´b` have a fresh flexible variable `'c` instead of `'a`, but the result of copying the inner occurrence of `´b` must be different, as its own occurrences of `'a` are now bound by the inner `'a` quantification. In other words, the result

- should be `< n : 'c; p : <m : 'a. < n : 'a; p :  'p > > as 'p >`
- and not `< n : 'c ; p : < m : 'a . 'b >> as 'b`

Do you agree? If this is right (if you believe that it captures the essence of why `copy_sep` is needed), it does sound like an example worth explaining in a comment, so that people understand the need to un-share types during instantiation. I could send a PR for this.

garrigue commented:
Yes, this is the essence of `copy_sep`.
But you should be careful that there is a need to distinguish between the `f` case and the `h` case, i.e. you cannot assume that there will be no univars left eventually, so you have to check wether you can close the recursion without changing the binding structure.
I'm still thinking how to simplify the code using the invariant that the only non-bound variables were those bound just outside.

garrigue committed:
simplify code by computing unbound univars

garrigue committed:
remove unused TypeSet.of_list

garrigue committed:
sharing

Octachron commented on testsuite/tests/typing-poly/poly.ml:
```diff
@@ -1516,30 +1516,33 @@ val n : < m : 'x 'a. ([< `Foo of 'x ] as 'a) -> 'x > = <obj>
+             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The method m has type 'a -> 'b but is expected to have type
+         [< `Foo of 'x ] -> 'c
+       The universal variable 'x would escape its scope
```
Note this example is still marked as `ok` even if it fails now.

This seems like a slight regression, however the slightly modified version 
```ocaml
object method m: 'x. [< `Foo of 'x] -> 'x = fun (`Foo x) -> x end
```
was already failing, thus this seems like a pre-existing issue.

garrigue replied:
Indeed, this code is basically wrong. A univar should not appear in a type whose row variable is bound at a more external position. This is correctly enforced when the row variable is also a univar, but not when it is a normal variable, ending up accepting meaningless types that cannot really be used (trying to call this method would fail, IIRC).

garrigue committed:
simplify using assumption that we are only opening a Tpoly at a time

garrigue commented:
Commit fba9003d39f7fd1f07c2299991de4985486c1398 simplifies the code by using the assumption`copy_sep` is only called on types whose most external `Tpoly` was removed. This basically mean that sharing can only happen in two ways. Either it goes through that external `Tpoly`, but then the type has no free univars, and `copy` is called; or it is an internal sharing, but then the structure of the original type is copied as is, replacing just those external univars with the fresh variables generated in `instance_poly'`, and we can keep the sharing inside that copy. As a result, there is no need to track the bound variables.

garrigue committed:
remove unused ~bound

garrigue committed:
need to share everything for the invariant to be correct

garrigue committed:
rename ~shared to ~visited

garrigue commented:
Note that this invariant is only valid if we really keep the sharing. I.e., `Tpoly` nodes should be shared, otherwise we may end up sharing nodes whose parent use different copies of the univars. The new version of `copy_sep` should look much simpler now; the so-called unsharing is only between the parts that contains free univars, and is copied using a hashtable to keep internal sharing, and the part that doesn't contain free univars, which uses `copy`.

gasche commented:
Independently of your last simplification, I thought about this during the day. I think that the core idea is that we are implementing a substitution function on univars, and that two occurrences of the same type node in different univar scopes will (naturally) be substituted differently. So instead of memoizing this type-graph function by keeping one result per node, we should keep one result per node *and* substitution argument (or free-univar context, the two are in direct correspondence). This is exactly the amount of unsharing that is necessary.

If this reasoning is right, then `copy_sep` should behave exactly as `copy` not only in the closed univar context (no univar substitution at all), but also within each slice of the typegraph for a given context. In particular:
1. We could consider implementing univar-substitutions in `copy` directly, instead of having two separate functions (but: this makes `copy` more complex as we cannot use the in-place memoization strategy for nonempty univar substitutions).
2. At least the code handling `Tvariant` and `Tfield` should be very similar in both functions. (When can a row variable be a univar?)

gasche commented on typing/ctype.ml:
```diff
@@ -1305,67 +1300,56 @@ let instance_class params cty =
+      Tarrow _ | Ttuple _ | Tvariant _ | Tconstr _ | Tobject _ | Tpackage _ ->
+        TypeHash.add visited ty t
+    | Tvar _ | Tfield _ | Tnil | Tpoly _ | Tunivar _ ->
+        ()
```
why are all these types excluded from being added to `visited`?

garrigue replied:
This was a left-over of the previous code; we didn't want to create recursion in a strange place. Now that we are sharing everything, there is no point of doing this selective cacheing, so I removed it, and just add everything.
By the way, the invariants guarantee that `Tvar`, `Tunivar` and `Tnil` cannot occur here anyway. And I believe that `Tfield` and `Tpoly` cannot be shared without sharing their parent, so this shouldn't make a difference.

gasche commented:
Remark: when we call `copy_sep` from instance_poly, it is my understanding that we know that the set of free univars is exactly the set of variables of the `Toply` node we opened. Whenever we encounter a `Tpoly` node, all its univars are bound in its body, so we can remove those univars from the current set of free univars. This would let us implement the logic here without computing the free univars for the traversed nodes of the type graph. (The `compute_univars` implementation is reasonably efficient, but it still traverses all nodes of the graph, including those without free univars and those at non-generic levels.)

lpw25 commented:
Re. `copy_sep`. IIRC this function disappears if we adopt the level-based approach to handling univars that @trefis was working on. Perhaps we should consider trying to get that merged rather than simplifying the existing approach. It would simplify quite a few parts of the system. There was also an initial attempt to implement something along the right lines in #10787.

gasche commented:
This discussion reminded me of univars vs. levels as well, and I agree that implementing @trefis' approach would be a promising medium-term approach. But I don't want to embark on this now personally (I'm trying to work a bit less on the compiler, not more). On the other hand, clarifying the current codebase is a more local transformation, and it seems useful to fix a bug in this case; I can't say no to that.

garrigue committed:
Add comment explaining how copy_sep works

garrigue committed:
just share everything

garrigue commented:
> If this reasoning is right, then `copy_sep` should behave exactly as `copy` not only in the closed univar context (no univar substitution at all), but also within each slice of the typegraph for a given context. In particular:
> 
> 1. We could consider implementing univar-substitutions in `copy` directly, instead of having two separate functions (but: this makes `copy` more complex as we cannot use the in-place memoization strategy for nonempty univar substitutions).
> 2. At least the code handling `Tvariant` and `Tfield` should be very similar in both functions. (When can a row variable be a univar?)

Actually, now that we are keeping the sharing, `copy_sep` does behaves exactly as `copy`. It is just much simpler because it uses a hahstable rather than complicated redirections.
It might be worth considering reimplementing `copy` and friends using the same approach, and seeing wether it is really more costly.
The idea of doing in-place memoization with multiple contexts seems overly complex to me.
Also performance is not a serious problem in `copy_sep`, since it is much less used than `copy`. Basically, only once by method call. Also on some explicitly annotated let constructs.

garrigue commented:
> Remark: when we call `copy_sep` from instance_poly, it is my understanding that we know that the set of free univars is exactly the set of variables of the `Toply` node we opened. Whenever we encounter a `Tpoly` node, all its univars are bound in its body, so we can remove those univars from the current set of free univars. This would let us implement the logic here without computing the free univars for the traversed nodes of the type graph. (The `compute_univars` implementation is reasonably efficient, but it still traverses all nodes of the graph, including those without free univars and those at non-generic levels.)

I have been thinking of that, and this doesn't work. The point is that `copy_sep` needs to copy its target as long as it contains free univars, independently of its level. And of course we need to revert to the normal behavior a soon as it does not contain free univars. For this we need to know the free univars in each node.

garrigue commented:
> Re. `copy_sep`. IIRC this function disappears if we adopt the level-based approach to handling univars that @trefis was working on. Perhaps we should consider trying to get that merged rather than simplifying the existing approach. It would simplify quite a few parts of the system. There was also an initial attempt to implement something along the right lines in #10787.

This PR is not just a readability improvement, but also fixes a soundness bug.
And I believe that the simplification is actually needed to fully fix the bug. See `g3` in the testsuite for why unsharing as it was done is incompatible with the fix.

garrigue committed:
use copy_rec

gasche commented on typing/ctype.ml:
```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
+   bound when under this [Tpoly] node, which has no free univars, and as
+   such is not part of the separate copy. In turn, this allows the separate
+   copy to keep the sharing of the original type without breaking its
+   binding structure.
```
I don't follow what is written in this `Assumption` paragraph.

> in the first call to [copy_sep], all the free univars should
   be bound by the same [Tpoly] node. This guarantees that they are only
   bound when under this [Tpoly] node

I agree.

>  they are only
   bound when under this [Tpoly] node, which has no free univars, and as
   such is not part of the separate copy

I believe that you say that all *bound* occurrences of these univars are dominated (in our traversal) by a Tpoly node that binds all these univars simultaneously, and never a strict subset of them, and so they are univar-closed subterms. This means that we don't need to keep track of arbitrary many possible free-univar substitutions (which would all need to be memoized separately), the only two substitutions are "the initial substitution of free univars we are instantiating by flexible variables" (memoized by `visited`) and "the empty substitution on univar-closed terms" (memoized inside `copy`).

(Note: traversing another `Tpoly` node that binds separate univars does not extend the substitution, so it does not require yet another separate memoization cache.)

You mentioned in the discussion that you assume that univars have unique binding sites, or at least the univars bound by the Tpoly node we are instantiating. But I think we are in fact only assuming that all Tpoly nodes that bind one of the free univars also bind all other free univars in the initial set. So we could in fact have duplicate copies of this initial Tpoly node without breaking the implementation choice.

>  In turn, this allows the separate copy to keep the sharing of the original type without breaking its binding structure.

I don't know what this means.

garrigue replied:
The story about using only 2 substitutions is correct. This is what this comment is supposed to say (in terms instantiation rather than substitution, but this means the same thing).

However, it is not true that traversing a [Tpoly] node that binds separate univars does not extend the substitution. If this [Tpoly] is not univar-closed, we need to copy it, and as result we also need to copy its univars. I.e. we extend the substitution with mapping from these univars to fresh univars. So when we duplicate a [Tpoly] node because it contains free univars, we also need to duplicate its whole skeleton (i.e. all the occurrences of the univars bound bind this [Tpoly]), which in turn may force use to duplicate more [Tpoly]'s... This is all fine because we duplicate as a graph.

The last sentence says that, despite the fact the separate copy may contain recursions on `Tpoly`'s, since the only really intantiated univars (to normal variables) are the ones that were given originally, and they are all bound by te same [Tpoly], this separate copy preserves the binding of univars of the original type. Maybe the reason you couldn't understand is just that you didn't see that we were copy all the univars, not just the instantiated ones.

gasche replied:
Thanks! I understand better now, but the last sentence of your comment is still obscure in my opinion -- I don't think that other people would necessarily understand what you mean.

I would like to understand this algorithm as a copy function with several distinct memoization tables (that let us preserve sharing and provides termination on cycles), and a natural/correct/clear criterion for when to move from one memo table to another. I understand the logic in the current implementation, which is to use two memo tables, one in the traversal contexts where the original-Tpoly-univars are free and another when they are bound. But why is this correct?

My original intuition is that we use a different memo-table for each substitution (from a list of free univars to types). But as you just pointed out, this is not quite right: when we go under a Tpoly node, we extend the current substitution (or, in the case of recursive occurrences of the original Tpoly, we shadow the substitution), so we get a different substitution. But we don't use a different memo table at this point (unless we shadowed the original univars). You speak of "the really instantiated univars" as the memoization key (in addition to the node), so to speak, but again, is there a simple explanation for why this is correct?

(My attempt at reasoning this out below.)

---

If we have a substitution ` rho` and another `rho'` that extends `rho` (in particular, it coincides with `rho` on all univars bound by `rho'`), it is correct to share the same memo-table between the two, under the assumption that all free univars in the substituted term are bound in `rho` (this is the natural extension of your "no other free univar at the beginning" assumption). Indeed, for the nodes that were substituted with `rho`, the univars of `rho'` were not free, so it would be equivalent to substitute them with `rho'`, and so we can consider that all nodes are substituted with `rho'` and may share a single memo-table.

But then we would be in trouble if there was another extension `rho''` of `rho`, in another part of the type graph, that was incompatible with `rho'` because it would bind some of the same univars to different values. This corresponds to a case of two distinct `Tpoly` nodes binding the same univars. Can this happen? (It would probably be fine in this case to share the substitutions of the common univars between the two occurrences, guaranteeing that `rho'` and `rho''` are in fact compatible.)

So as things stands, my explanation would be as follows:
- let's call "initial univars" the univars of the instantiated Tpoly nodee and "initial substitution" the corresponding substitution
- whenever we go under a Tpoly node that does not shadow the initial univars, we can extend the initial substitution in a way that is compatible with all other extensions
- therefore we can use a single memo-table for the initial substitution and all its extensions
- we assume that the only Tpoly nodes that shadow some initial univars is the initial Tpoly node: (1) it shadows all of them at once, and (2) no other univars (from any extension) is free under immediately it
- therefore, when encountering recursive occurrences of the initial Tpoly node, we have an incompatible substitution on the initial univars only; this requires a separate memo-table, which is done by `copy`


---
Question: if I read the code correctly, `copy` does not freshen the univars bound by the `Tpoly` node it encounters (`copy` itself has no Tpoly-specific logic, and `Btype.copy_type_desc` keeps univars unchanged). In particular, separate copies will keep using the same univars, we are not trying to preserve unicity of bound variables. Why can't we use this strategy for `copy_sep`, keep unchanged the univars of the non-initial Tpoly node we encounter during the traversal? Then we would really have only two substitutions, so the explanation would be quite a bit simpler.

garrigue replied:
For the reasoning, this is indeed what I had in mind. I agree that the last sentence is too short to express what you explained in 20 lines, but should we really put proofs (even informal) in the code, and if not there, where ?

For the question, there is a misunderstanding, as for `Tunivar` as for `Tvar` the identity of the type is that of the node. So even if `copy_type_desc` is the identity we do get a fresh type node.
This said, I have also simplified `copy_sep`, to remove the `Tpoly` case that appears to be no longer necessary. As a side-effect, this removes the reliance on the invariant that the same `Tunivar` cannot be bound by physically distinct (but semantically identical) `Tpoly`.

gasche replied:
Nice simplification.

I think that the simplified version supports my intuition that we are not extending the substitution when encountering Tpoly node whose univars are disjoint from the domain. In a paper definition we would write something like

```
x \notin \rho
--------------------------------
(lam x. t)[rho] := lam x.t[rho]
```

Note that the `x` in the result is a different binder from the `x` in the input, so we implicitly freshen as in `copy_sep`. We could write `[rho, x/x]` on the right-hand-side but it is not necessary.

So we could justify this function with a model with only two distinct substitutions, without considering extensions.

garrigue replied:
That works only if you assume that we track scoping properly. Currently, we rather follow the Barendregt convention: bound type variables occurring in different places must be different. This makes reasoning easier.

gasche replied:
I don't agree/understand. My point is that we perform a substitution on some univars, and a *copy* (in a memo-table that is indexed by the substitution) on all other nodes, including the other univars (so: no extension of substitutions). There are two different subtitutions, the instantiating substitution on the free occurrences of the initial univars, and the empty substitution for subterms that do not contain free occurrences of the initial univars (including under the recursive occurrences of the initial Tpoly node).

This sounds like a simple specification of this function (much simpler than considering compatible extensions etc.), and it does clearly correspond to the code now that you have removed the specific Tpoly handling.

gasche replied:
> should we really put proofs (even informal) in the code, and if not there, where ?

To understand functions on OCaml types, I need two things:

1. A high-level specification of what the function does, as an operation on type-graphs, described as declaratively as possible. (This high-level operation should make sense to me.)
2. An explanation of how that operation is realized in the implementation I am about to read. (The relation between the abstract operation and the implementation choices should make sense to me.)

The high-level operation is a substitution with substitution-indexed memo tables. (The memo table matter in the high-level specification as they are necessary for termination, even if we don't care too much about preserving sharing in general.) The discussion we are having (just two substitutions, or their many extensions under additional free univars) is about (1) how to define/present this high-level operation. I understand both choices of presentation. But one makes (2) very clear (there are two substitutions in the abstract, so two memo table, and I see how they are realized in the implementaiton), and the other (with extensions) makes (2) much less clear, because I would then naturally expect the implementation to have many memo-tables. In this second case, to understand (2), the relation between the specification and the implementation, I would indeed need an informal correctness argument for why many memo-tables can be condensed in just two. This is just about making sense of the code (and its relation to the specification), not a general belief that we should always informally prove correctness.

garrigue replied:
Should we schedule another online discussion?
I have been spending more time writing explanations here than fixing the bug. If I need to write a paper on the subject before we fix the bug, this all becomes extremely time consuming, and I will end up not fixing bugs.

I'm not sure of what would qualify as high-level specification for you.
I repeat what this function is doing at the graph level:
* it first produces a copy of the part of the graph reachable from the root that contain free univars; in this copy the univars we are instantiating are replaced with normal variables, the other ones are replaced by fresh univars (because when you copy a `Tunivar` you are creating a fresh one). This copying ignores levels, as it is required to preserve the binding structure.
* it then connects this copy with another instance of the graph corresponding to parts that do not contain free univars, and as such can use the normal `copy` function, whose behavior depends on levels.
* there are some subtle cases for objects and variants, but they are better understood separately.

garrigue replied:
OK, let me try to give a theoretical specification. This will ignore objects and polymorphic variants for now.

Let's start with trees rather than graphs.
The idea is indeed that we are copying with a substitution for univars. This substitution can return either a normal variable or a fresh univar. I write `'A` for univars.
```
t[rho] = t       if FU(t) = emptyset
'A[rho] = rho('A)
Tpoly(t, 'A1...'An)[rho] = Tpoly(t[rho,'B1...'Bn/'A1...'An], 'B1...'Bn)
other cases are stuctural
```
where `FU(t)` stands for the free univars of `t`.
We can simulate the copying on the remainder by creating the fresh nodes at generic level and taking an instance of `t[rho]` with `copy`, so that part is orthogonal.

Now, of course this specification does not work on cyclic graphs. When can we share the result of substitution?
Basically we can have `t[rho] = t[rho']` if `rho|FU(t) = rho'|FU(t)`.
Without any assumption, this can be tricky, and gave the criterion in the original code.
But if we assume that the original `rho` contains only univars bound in the same `Tpoly`, and that any univar is only bound by one `Tpoly` node, then the scoping rules tell us that whenever we apply the substitution on the same node, we have indeed `rho|FU(t) = rho'|FU(t)`.
* the assumption tells us that the root `Tpoly` node has no free univars, so that `t_root[rho] = t_root[rho'] = t_root` by the first rule. This also means that we never overwrite the univars of the original `rho`.
* for other nodes, when we reach a shared node, scoping tells us that the only univars that can be free in it are those that were bound on all the branches that lead to it, or those in the original `rho`. But then, since we assumed that a univar is only bound by a single `Tpoly` node, this means that this node was already shared, and that all the univars it binds are mapped to the same fresh univars in `rho` and `rho'`. Similarly, those from the original `rho` always map to the same variable. Hence `rho|FU(t) = rho'|FU(t)`, and we can share the result of substituting this node.

So it is fine to preserve all sharing. And once we preserve all sharing, the polymorphic variant and object cases become reasonably easy.

gasche commented on typing/ctype.ml:
```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
+          (* In that case we should keep the original, but we still
+             call copy to correct the levels *)
+          if keep then (add_delayed_copy t copy_scope ty; Tvar None) else
           let more' = copy_rec ~may_share:false more in
```
What is `may_share` doing now, why do we need `may_share:false` here and in the `Tfield` case?

garrigue replied:
`may_share` is here to ensure that when we copy an object or a polymorphic variant, we also copy its row. It preserves the invariant that two types with the same row are identical.

gasche commented on typing/ctype.ml:
```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
+             call copy to correct the levels *)
+          if keep then (add_delayed_copy t copy_scope ty; Tvar None) else
           let more' = copy_rec ~may_share:false more in
           let fixed' = fixed && (is_Tvar more || is_Tunivar more) in
```
In the `copy` version, the call to `copy_row` always use `fixed = true`, here the definition of `fixed'` is fairly different. Why? Relatedly, can you explain what the `fixed` parameter of `copy_sep` is meant to do? (I see that `instance_poly` and `instance_label` sometimes pass `true`, but I can't spot the logic there.)

garrigue replied:
A polymorphic variant type whose row variable is a univar should be _fixed_, i.e. one cannot add or remove row fields through unification, and `Reither` can only contain one type. When we instantiate this univar into a normal one, the row is no longer fixed. The meaning of the `fixed` parameter of `copy_row` is to keep the fixedness information in the row. When it is false, the copy is no longer fixed.

gasche commented on typing/ctype.ml:
```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
+             call copy to correct the levels *)
+          if keep then (add_delayed_copy t copy_scope ty; Tvar None) else
           let more' = copy_rec ~may_share:false more in
           let fixed' = fixed && (is_Tvar more || is_Tunivar more) in
```
In the `copy` version there is one case related to GADTs where the `fixed` field of the copied row is in fact changed:

https://github.com/ocaml/ocaml/blob/88c3fc65774ec669d9ed8e6ad3ef136ba0594a40/typing/ctype.ml#L1089-L1095

do we need similar logic here?

garrigue replied:
As indicated, the reason is explained in #6163.
Namely, if `row_more`is a `Tconstr`, the row should also be fixed. Apparently this is not properly enforced by reification. This should probably rather be done inside reification than inside `copy` but this is clearly orthogonal to this PR.

gasche replied:
My question was whether the same issue could occur within the input of `copy_sep`, instead of `copy`, as so we would need to patch fixedness in the same way for correctness. It is not clear to me that `copy_sep` will never encounter a Tconstr row variable with incorrect fixedness, but if it is to you then maybe that's good enough.

garrigue replied:
As said above, if there is a need to fix that, this should not be in `copy_sep` but in the reification code.
Should try to come up with some example.

gasche replied:
My understanding of this point:

- currently this case in `copy` fixes a bug, but you believe the fix is not appropriate and should be done somewhere else (I agree with this intuition)
- by not reproducing the fix in `copy_sep`, we expose ourselves to a similar bug
- but you prefer to not duplicate the non-appropriate fix, and wait for later to do a better fix that would require no code in `copy` or `copy_sep`

That's okay with me, with the understanding that we decided to leave this bug open and fix it later.

garrigue replied:
Yes, this is my view. With the recurring problem that, in order to fix a bug properly, you need a counter-example, and in this case it is not so easy to come up with `copy_sep`. Fixing reification should be easier to test.

gasche reviewed (commented):
The new iteration of the code is indeed much simpler, and thanks a lot for writing a comment at the beginning of the function along the lines we discussed. (I had some discussion on the last paragraph of this header comment, see inline messages.)

Now I am satisfied that I understand what `copy_sep` is doing, and I think it's time we start discussing your actual bugfix :-)

gasche commented:
While looking at users of `copy_sep`, I found two places with comments about `copy_sep` that I don't understand:

1. In `instance_label`

https://github.com/ocaml/ocaml/blob/88c3fc65774ec669d9ed8e6ad3ef136ba0594a40/typing/ctype.ml#L1405-L1406

What does this mean? As far as I can tell, the only `Tsubst` nodes introduced by `instance_poly'` (that is, `copy_sep`) are those introducd by the internal calls to `copy` itself, so why would we need to postprocess the result with `copy` again? (Or is this about the Tlink nodes introduced by  `add_delayed_copy` after the copies have been forced?)

2. In `check_univars`

https://github.com/ocaml/ocaml/blob/88c3fc65774ec669d9ed8e6ad3ef136ba0594a40/typing/typecore.ml#L2494-L2498

do I correctly understand that `Tvar` is a typo in this comment, and `Tunivar` is meant? On non-generic types, `copy_sep` only copies the spine from the root of the type to all free occurrences of a univar.

garrigue commented:
> While looking at users of `copy_sep`, I found two places with comments about `copy_sep` that I don't understand:
> 
> 1. In `instance_label`
> 
> https://github.com/ocaml/ocaml/blob/88c3fc65774ec669d9ed8e6ad3ef136ba0594a40/typing/ctype.ml#L1405-L1406
> 
> What does this mean? As far as I can tell, the only `Tsubst` nodes introduced by `instance_poly'` (that is, `copy_sep`) are those introducd by the internal calls to `copy` itself, so why would we need to postprocess the result with `copy` again? (Or is this about the Tlink nodes introduced by `add_delayed_copy` after the copies have been forced?)

The problem is that calling [copy] on a type introduces [Tsubst] nodes inside it. [copy_sep] needs to work on the original type, so it should be called on a type for which [copy] has not been used.

> 2. In `check_univars`
> 
> https://github.com/ocaml/ocaml/blob/88c3fc65774ec669d9ed8e6ad3ef136ba0594a40/typing/typecore.ml#L2494-L2498
> 
> do I correctly understand that `Tvar` is a typo in this comment, and `Tunivar` is meant? On non-generic types, `copy_sep` only copies the spine from the root of the type to all free occurrences of a univar.

Indeed, that looks like a typo.

gasche commented on typing/ctype.ml:
```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
-            copy_sep ~copy_scope ~fixed ~free ~bound ~may_share:true
-              visited t1 in
-          Tpoly (body, tl')
+          List.iter2 (TypeHash.add visited) tl tl';
```
Do we know for sure that the `tl` are not already in `visited` at this point? If yes, could we consider adding an assertion? If no, is the code correct when that happens?

garrigue replied:
I think this is true, but this invariant is not required. Even if the `Tpoly` nodes are not physically identical, as long as they share the same univars they should be semantically identical, and the code is still fine.

gasche commented:
https://github.com/ocaml/ocaml/blob/88c3fc65774ec669d9ed8e6ad3ef136ba0594a40/typing/ctype.ml#L1396-L1408

> The problem is that calling [copy] on a type introduces [Tsubst] nodes inside it. [copy_sep] needs to work on the original type, so it should be called on a type for which [copy] has not been used.

Ah, I understand now.

Thinking out loud: we could also use the invariant that, for a given `copy_scope`, Tlink and Tsubst are only created for memoizing univar-closed subtypees. Then we could extend `copy_sep` on `Tsubst` nodes exactly as `copy`, and I think that then we could copy the two parts in any order there, and maybe we could even get rid of the `add_delayed_copy` mechanism and use `copy` directly?

gasche commented on typing/ctype.ml:
```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
           let keep = is_Tvar more && get_level more <> generic_level in
+          (* In that case we should keep the original, but we still
+             call copy to correct the levels *)
+          if keep then (add_delayed_copy t copy_scope ty; Tvar None) else
```
`keep` being true tells us that the polymorphic variant has a non-generic row variable. If I understand the `add_delayed_copy` here, the idea is to `copy` the whole polyrmorphic variant in this case. Why is this correct? What if one of the variant fields contains one of the instantiated univars?

This is the key of the bugfix in this PR. I understand that you want to guarantee that if the Tvariant is not shared, then the row is also not shared. Here you do this by always sharing the Tvariant when you know that you need to share the row.

Instead I would expect the current `may_share` logic to be enough here. At this point we know that we will unshare the `Tvariant` node, so we want the row variable to be unshared as well. Why isn't `may_share:false` enough? Because the check for sharing nodes in `copy_sep` is `is_Tvar ty || may_share && is_empty univars`, so Tvar rows are shared even when `may_share;false`. Would removing the `is_Tvar ty` part of the condition fix the issue? (Obviously this would not make any difference when `may_share:true`, as Tvar nodes are univar-free.)

garrigue replied:
Actually, this is not the main part of the fix. The main part is the switch from using `visited` only for recursion (and substitution of univars) to using it to keep sharing.
Now, for this specific case. If `row_more` is a non-generic type variable, then it means that structure of the row should be shared with the original. The safest way to do that is to keep the original one.
The previous code duplicated the row, but kept the sharing in the `row_field_desc`'s. This had the strange consequence that some function could be typable, but not applicable because as soon as you tried to unify it with something else as a univar would leak. I came to the conclusion that it was better to keep the original row than to try to create this kind of copy, in particular this avoid delaying errors.
Note that this mechanism of copying while keeping sharing with the original is also used in `Subst.typexp`. There the problem is a bit different because one no longer refers to the original after copying, so there `keep` acts more like a safety mechanism.

gasche replied:
I still wonder whether this is correct: could we have a variant with a non-generic row variable, but fields that refer to the instantiated univars? In this case, isn't the `keep=true` case iconrrect?

garrigue replied:
This is indeed the case, and there are such examples in the test suite. The point is that we are going to fail immediately (when we check the scoping of univars on the result of the substitution), rather than later on as was the case before.

garrigue replied:
Stated differently, a type whose row variable is a Tvar should not contain free univars to start with. Currently this is not enforced explicitly (this should probably be changed, but this concerns a different part of the code); we cannot fail inside `copy_sep` so we delay the failure until just after. Note also that if the row variable is a Tunivar, then all free univars should be bound outside of the scope of this Tunivar; this one is properly enforced.

gasche replied:
I don't understand, probably because my mental model of variants / row variables is wrong. Intuitively I see nothing wrong with having a non-generic row variable and univars in the field types. (We allow mixing non-generic *type* variables and univars, why should row variables be handled differently?) Does this assumption that they cannot happen come from the OCaml syntax for rows, where we cannot name the row variable itself and have to name the whole variant instead, capturing the field types as the same time?

That is: I think that

```
let f (x : [ | 'r]) (y : < m : 'b . [ `Foo of 'b | 'r ] >) = ignore y#m
```

should be accepted, but I guess we currently have to write this as

```
let f (x : 'a) (y : < m : 'b . ([> `Foo of 'b ] as 'a) >) = ignore y#m
```

and this is invalid / can never happen because `'b` escapes through `'a`. Is this the idea?

(But: aren't there ways to get a non-generic row type variable without naming/escaping the field types?)

gasche replied:
Note: I still wonder why we have the `is_Tvar ||` condition for the copy case. Why is it necessary?

garrigue replied:
First, about why it is wrong to duplicate the fields of a row whose (normal) variable is not generic.
There is actually a mismatch between the implementation and the theory. The theory is based on Ohori's kinds, where a row variable is seen as representing the whole variant type, rather than just its extension as in Remy's approach, which is followed in the implementation. From this point of view, having a normal row variable capture a univar in one of the fields is incorrect in the theory. Even if we follow Remy's approach, there is a problem with the `Reither` fields. Those fields can be extended with extra types, that may eventually be unified together.  Sharing the row variable means also sharing the extension variable of all the `Reither` fields. If this field contains univars that are going to be instantiated, then it means that we end up someday unifying the extension with both the instantiated univars (which are normal vars) and the original ones, which will inevitably cause a clash, and a harder to understand one.
You can see this in the example `j` in the testsuite:
```ocaml
let j (x : < m : 'a. [< `A of 'a ] -> 'c > as 'c) = x#m;;
```
It used to be accepted, but you will have a hard time defining an object with this type, and even if you succeed, unification will fail when try to pass it to `j` or to use its result. The reason is that inside the type of `x#m` there are two instances `[< A of 'a]` sharing the same extension. This should be something like:
```ocaml
[< `A of 'b & 'e] -> (< m : 'a. [< `A of 'a & 'e] -> 'c > as 'c)
```
You see the contradiction between the two instances, as using the field `A` would result in unifying `'b` and `'a`, which is clearly impossible.
If we follow Remy's approach, it would be OK to allow univars in the `Rpresent` fields, but then we need to build the corresponding theory, and I think it is intrinsically more complex, for limited practical benefits. The reason we are still sticking with implementation is more related to the principality tracking: by using a different level for the `Tvariant` and `row_more` we can generalize the level of more nodes, i.e. see more types as known, which is useful in practice.

Now, about the `is_Tvar`. If the row variable is either `Tunivar` or `Tconstr`, it is OK to duplicate the row. Its contents are _fixed_ in the original, so that there is no need to share extension between the original and the copy. This is for instance what is happening in example `g`, where we simultaneously instantiate `'a` and the row variable `'b`:
```ocaml
let g (x : < m : 'a 'b. <n : ([< `A of 'a ] as 'b) -> 'c > > as 'c) = x#m;;
```

gasche replied:
Thanks for the explanation regarding the rows. I hadn't realized that the row variable in OCaml is supposed to represent the whole variant/object. (But then this morning I was tearing my hair trying to understand `Tobject` code which I guess makes more sense now.)

>  Even if we follow Remy's approach, there is a problem with the `Reither` fields. Those fields can be extended with extra types, that may eventually be unified together. Sharing the row variable means also sharing the extension variable of all the `Reither` fields.

I thought, when reading the code of `copy` and `copy_sep`, that doing this properly was the job of the `keep` parameter to `copy_row` -- that it would take into account, during field copy, the fact that the row variable is non-generic and will be shared, so that the field extensions are shared as well.

> Now, about the `is_Tvar`. If the row variable is either `Tunivar` or `Tconstr`, it is OK to duplicate the row. Its contents are fixed in the original, so that there is no need to share extension between the original and the copy.

I'm not sure you are replying to the question I was asking. I understand your answer as related to this occurrence of `is_Tvar`:

https://github.com/ocaml/ocaml/blob/f669914c672c8b77f24e96d81c6b7100039086ee/typing/ctype.ml#L1350

but I was meaning to ask about this one:

https://github.com/ocaml/ocaml/blob/f669914c672c8b77f24e96d81c6b7100039086ee/typing/ctype.ml#L1334

This occurrence of `is_Tvar` forces copying the row variable even when `may_share:false` is passed to the recursive calls of `copy_sep` in the Tvariant and Tfield case. I wonder why this is the correct thing to do. I may be missing something, but I have the impression that if we didn't have this `is_Tvar ty ||` condition we could safely use the same code in the Tvariant case whether `keep` is true or not (... assuming that `keep` is in charge of keeping/sharing the field extensions when the row variable is not generic), just like `copy` is doing. (And `copy` doesn't have an `is_Tvar` condition to decide whether to share or to duplicate.)

garrigue replied:
Ugh. Now you're talking about 20 year old code...

Originally, `copy_sep` was not preserving sharing, so it is clear that one had to call `copy` on it to preserve sharing.
Now that `copy_sep` preserves sharing, is it still needed ?
At least it is needed for objects, if the row variable is not generic. In that case we allow copying  fields using `copy_sep`, but the row variable, which allows to add extra fields, should still be shared with the original.
For variants, it can only be called if the row variable is generic, so that the result of copying it with `copy_sep` or `copy` should be identical.

garrigue replied:
As for `keep`, while it is sound, it does not always make sense to share an extension between incompatible types.
More precisely, `keep` works well for `Subst.typexp`, as it is there used to compute signatures of modules: the original type is no longer used after the substitution, so the compatibility doesn't matter. However, in `copy_sep` where we might want to unify both the original and the copy later on, this is essentially delaying problems, and I prefer to stick to a well-understood theory, which requires correct scoping from the beginning.

garrigue commented:
> Thinking out loud: we could also use the invariant that, for a given `copy_scope`, Tlink and Tsubst are only created for memoizing univar-closed subtypees. Then we could extend `copy_sep` on `Tsubst` nodes exactly as `copy`, and I think that then we could copy the two parts in any order there, and maybe we could even get rid of the `add_delayed_copy` mechanism and use `copy` directly?

I don't think it would work. Again, `copy_sep` needs to refer to the original type, which is temporarily lost once it is replaced by `Tsubst`. We could of course extend `Tsubst` to keep a reference to the original type, but this introduces extra complexity.

garrigue committed:
use copy_sep for Tunivar too

garrigue committed:
simplify further

gasche reviewed (approved):
Approved. We had a long discussion of this function, and while I still don't understand all the details I have a good enough understanding to agree that the fix makes sense. In addition, the code is much, much simpler now, and there is some documentation of what the function is doing. Very nice!

gasche commented:
(Also: thanks to @garrigue we have a very good testsuite coverage of this issue now, and in general of examples mixing univars and variants or object types.)
